ores/scoring_context.py

```diff
@@ -0,0 +1,224 @@
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        return score
+
+    def _generate_root_datasources(self, model_names):
+        for model_name in model_names:
+            for dependency in dependencies.dig(self.model_features(model_name)):
+                if isinstance(dependency, Datasource):
+                    yield dependency
+
+    def extract_root_dependency_caches(
+            self, model_names, rev_ids, injection_caches=None):
+        """
+        Extracts a mapping of root datasources capable of generating the
+        features needed for a particular set of models.
+
+        :Parameters:
+            model_names : `list` ( `str` )
+                The names of a :class:`~revscoring.ScorerModel` to
+                extract the roots dependencies for
+        """
+        # Make a copy of dependency_caches
+        root_caches = {}
+        for rev_id in rev_ids:
+            injection_cache = injection_caches.get(rev_id) \
+                              if injection_caches is not None else {}
+            root_caches[rev_id] = {key: value
+                                   for key, value in injection_cache.items()}
+
+        # Find our root datasources
+        root_datasources = \
+            list(set(self._generate_root_datasources(model_names)))
+
+        # Extract the root datasource and update the root_caches in-place
+        start = time.time()
+        error_root_vals = self.extractor.extract(
+            rev_ids, root_datasources, caches=root_caches)
+
+        # Check each extraction for errors
+        errors = {}
+        for rev_id, (error, root_vals) in zip(rev_ids, error_root_vals):
+            if error is not None:
+                errors[rev_id] = error
+                del root_caches[rev_id]
+        logger.debug("Extracted root datasources for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, set(model_names), rev_ids,
+                             round(time.time() - start, 3)))
+
+        # Note that root_caches should have been modified in place
+        return root_caches, errors
+
+    @classmethod
+    def from_config(cls, config, name, section_key="scoring_contexts"):
+        """
+        Expects:
+
+            scoring_contexts:
+                enwiki:
+                    scorer_models:
+                        damaging: enwiki_damaging_2014
+                        good-faith: enwiki_good-faith_2014
+                    extractor: enwiki
+                ptwiki:
+                    scorer_models:
+                        damaging: ptwiki_damaging_2014
+                        good-faith: ptwiki_good-faith_2014
+                    extractor: ptwiki
+
+            extractors:
+                enwiki_api: ...
+                ptwiki_api: ...
+
+            scorer_models:
+                enwiki_damaging_2014: ...
+                enwiki_good-faith_2014: ...
+        """
```

Comment:
Is there a formal grammar for specifying how the configuration should be? That would be more useful IMO than an (outdated?) example